Should <button> have `user-select: none;` by default?

karlcow opened the pull request:
#### c73eea082ab652d680f34de6924a4c9829aaa3ea
<pre>
Should &lt;button&gt; have `user-select: none;` by default?
<a href="https://bugs.webkit.org/show_bug.cgi?id=244448">https://bugs.webkit.org/show_bug.cgi?id=244448</a>
rdar://99262559

Reviewed by NOBODY (OOPS!).

In the section 6.1. Controlling content selection had this
recommendation for the UA stylesheet for HTML.
button, meter, progress, select { user-select: none; }
<a href="https://drafts.csswg.org/css-ui/#content-selection">https://drafts.csswg.org/css-ui/#content-selection</a>
seeAlso <a href="https://github.com/whatwg/html/issues/8228">https://github.com/whatwg/html/issues/8228</a>
This will need to be unprefixed once Bug 208677 is fixed
<a href="https://bugs.webkit.org/show_bug.cgi?id=208677">https://bugs.webkit.org/show_bug.cgi?id=208677</a>

* LayoutTests/accessibility/mac/search-predicate-from-ignored-element-expected.txt:
* LayoutTests/accessibility/mac/search-predicate-from-ignored-element.html:
* LayoutTests/accessibility/mac/textmarker-range-for-range-expected.txt:
* LayoutTests/editing/pasteboard/paste-noscript-expected.txt:
* LayoutTests/editing/pasteboard/paste-noscript-xhtml-expected.txt:
* Source/WebCore/css/html.css:
(button, meter, progress, select):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c73eea082ab652d680f34de6924a4c9829aaa3ea

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107645 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16696 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40540 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116792 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116196 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111537 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18118 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8014 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99816 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113399 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13674 "Found 2 new test failures: editing/pasteboard/paste-content-with-overflow-auto-parent-across-origin.html, fast/mediastream/mock-media-source-webaudio.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96875 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41346 "Found 1 new test failure: editing/pasteboard/paste-content-with-overflow-auto-parent-across-origin.html (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95588 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28505 "Found 1 new test failure: editing/pasteboard/paste-content-with-overflow-auto-parent-across-origin.html (failure)") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83190 "Found 3 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/replaced-objects, /WebKitGTK/TestDownloads:/webkit/Downloads/local-file-error (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9683 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29854 "Found 1 new test failure: editing/pasteboard/paste-content-with-overflow-auto-parent-across-origin.html (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10358 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6745 "Found 1 new test failure: editing/pasteboard/paste-content-with-overflow-auto-parent-across-origin.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15834 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49439 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11918 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9970b18514cfabfe6cc807d12f1123fd9068d842. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10017

nt1m commented on Source/WebCore/css/html.css:
```diff
@@ -404,6 +404,10 @@ select, button, meter, progress, input:is([type="button"], [type="submit"], [typ
 }
 
+button, meter, progress, select {
+    user-select: none;
```
WebKit hasn't unprefixed user-select yet.

```suggestion
    -webkit-user-select: none;
```

karlcow replied:
Thanks I have modified the css.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d17a60c6f9a5013a3e301b0ee4d894a3edb9ffd2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10017

karlcow commented:
There are a couple of tests to fix. that should be easy, but I'm worried about two tests related to accessibility. I wonder if it will change expectations for users. 

- [accessibility/mac/search-predicate-from-ignored-element.html](https://github.com/WebKit/WebKit/blob/main/LayoutTests/accessibility/mac/search-predicate-from-ignored-element.html) ([test history](https://results.webkit.org/?suite=layout-tests&test=accessibility/mac/search-predicate-from-ignored-element.html))
- [accessibility/mac/textmarker-range-for-range.html](https://github.com/WebKit/WebKit/blob/main/LayoutTests/accessibility/mac/textmarker-range-for-range.html) ([test history](https://results.webkit.org/?suite=layout-tests&test=accessibility/mac%20/textmarker-range-for-range.html))

cookiecrook commented:
Yes. I would not expect `user-select: none;` to inhibit either of those. Looping in @AndresGonzalezApple

karlcow committed:
Should <button> have `user-select: none;` by default?
https://bugs.webkit.org/show_bug.cgi?id=244448
rdar://99262559

Reviewed by NOBODY (OOPS!).

In the section 6.1. Controlling content selection had this
recommendation for the UA stylesheet for HTML.
button, meter, progress, select { user-select: none; }
https://drafts.csswg.org/css-ui/#content-selection
seeAlso https://github.com/whatwg/html/issues/8228
This will need to be unprefixed once Bug 208677 is fixed
https://bugs.webkit.org/show_bug.cgi?id=208677

* LayoutTests/accessibility/mac/search-predicate-from-ignored-element-expected.txt:
* LayoutTests/accessibility/mac/search-predicate-from-ignored-element.html:
* LayoutTests/accessibility/mac/textmarker-range-for-range-expected.txt:
* LayoutTests/editing/pasteboard/paste-noscript-expected.txt:
* LayoutTests/editing/pasteboard/paste-noscript-xhtml-expected.txt:
* Source/WebCore/css/html.css:
(button, meter, progress, select):

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/c73eea082ab652d680f34de6924a4c9829aaa3ea. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/10017

karlcow commented:
@AndresGonzalezApplefor the accessibility tests I have modified

cookiecrook commented:
I'm not sure I understand @karlcow's "👍" so I'm clarifying my prior statement, "I would not expect `user-select: none;` to inhibit either of those."

By that I mean, yes, it's odd that those are broken, which may indicate a sleeper accessibility problem with `user-select: none` that we haven't hit before. Please don't commit this change until @AndresGonzalezApple (or CF) has diagnosed and approved. Thanks.

karlcow commented:
> By that I mean, yes, it's odd that those are broken, which may indicate a sleeper accessibility problem with user-select: none that we haven't hit before. Please don't commit this change until @AndresGonzalezApple (or CF) has diagnosed and approved. 

Ha! Thanks for the clarification. That was exactly my worries.

AndresGonzalezApple commented:
This change will have an impact in accessibility support of these elements. So we will need to find an alternative for accessibility before going ahead with this change. To determine the text under an element, accessibility relies in the same machinery as editing, namely TextIterators. So if the text under a <button> cannot be extracted, say to copy to the clipboard, accessibility will not be able to retrieve the textual label either, thus making the button label-less for an accessibility client. Further investigation and possible implementation changes would be required to ensure accessibility can retrieve the text under an element with `user-select: none`.

pxlcoder commented:
> This change will have an impact in accessibility support of these elements. So we will need to find an alternative for accessibility before going ahead with this change. To determine the text under an element, accessibility relies in the same machinery as editing, namely TextIterators. So if the text under a cannot be extracted, say to copy to the clipboard, accessibility will not be able to retrieve the textual label either, thus making the button label-less for an accessibility client. Further investigation and possible implementation changes would be required to ensure accessibility can retrieve the text under an element with `user-select: none`.

@AndresGonzalezApple How does accessibility work for a `<div>` that contains text, with `user-select: none`. Is that also broken? Curious if there's a larger issue here.

pxlcoder commented:
Also note that authors could set `user-select: none` on `<button>` themselves, so it sounds to me like there is already an AX bug that needs to be fixed?

Ahmad-S792 commented:
@karlcow - should we carve out `button` and add remaining part of specification?

karlcow commented:
Let me close this. Not enough consensus on what to do here.
